src/parquet/UtilParquet.cpp

```diff
@@ -94,6 +122,67 @@ int cpp_getPrecision(const char* filename, const char* colname, char** errMsg) {
   }
 }
 
+static int getTypeOfFieldAtIdx(std::shared_ptr<arrow::Schema> sc,
+                               int idx) {
+  auto myType = sc -> field(idx) -> type();
+
+  if(myType->id() == arrow::Type::INT64)
+    return ARROWINT64;
+  else if(myType->id() == arrow::Type::INT32 || myType->id() == arrow::Type::INT16)
+    return ARROWINT32; // int16 is logical type, stored as int32
+  else if(myType->id() == arrow::Type::UINT64)
+    return ARROWUINT64;
+  else if(myType->id() == arrow::Type::UINT32 ||
+          myType->id() == arrow::Type::UINT16)
+    return ARROWUINT32; // uint16 is logical type, stored as uint32
+  else if(myType->id() == arrow::Type::TIMESTAMP)
+    return ARROWTIMESTAMP;
+  else if(myType->id() == arrow::Type::BOOL)
+    return ARROWBOOLEAN;
+  else if(myType->id() == arrow::Type::STRING ||
+          myType->id() == arrow::Type::BINARY ||
+          myType->id() == arrow::Type::LARGE_STRING)
+    return ARROWSTRING;
+  else if(myType->id() == arrow::Type::FLOAT)
+    return ARROWFLOAT;
+  else if(myType->id() == arrow::Type::DOUBLE)
+    return ARROWDOUBLE;
+  else if(myType->id() == arrow::Type::LIST)
+    return ARROWLIST;
+  else if(myType->id() == arrow::Type::DECIMAL)
+    return ARROWDECIMAL;
+  else {
+    return ARROWERROR;
+  }
+}
+
+int cpp_getAllTypes(const char* filename, int* types_out, char** errMsg) {
+  assert(types_out); // this has to be allocated by the caller
+
+  std::shared_ptr<arrow::Schema> sc;
+
+  try {
+    getSchema(filename, &sc, errMsg);
+  } catch (const std::exception& e) {
+    *errMsg = strdup(e.what());
+    return ARROWERROR;
+  }
+
+  const int num_cols = sc->num_fields();
+  for(int i=0 ; i<num_cols ; i++) {
+    int ret = getTypeOfFieldAtIdx(sc, i);
+    if (ret == ARROWERROR) {
+      std::string msg = "Unsupported type";
+      *errMsg = strdup(msg.c_str());
```

Comment:
strdup is unnecessary, `.c_str` returns new memory

Reply:
This is inherited code, so I didn't write anything specific knowingly here, but https://en.cppreference.com/w/cpp/string/basic_string/c_str seems to suggest that it does _not_ return new memory.

Reply:
I started putting together a `ParquetCall` context manager type where calls to C layer can be made within a `manage` statement and error is handling while exiting the context. That has sprawled way too much and I reverted back. I think we need to contain calls to C, error message/`ARROWERROR` propagation to a well-defined piece of code rather than sprinkling the Chapel module with it. I just updated the TODO list in the OP to include that.